src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
+        metadata[model.MD_PIXEL_SIZE] = (px, py, round(pz, ndigits=7))
+
+        # z coordinate of center
+        cx, cy = metadata[model.MD_POS]
+        cz = (self._zpos[0] + self._zpos[-1]) / 2
+        # upgrade the metadata with the z center position
+        metadata[model.MD_POS] = (cx, cy, round(cz, ndigits=5))
+
+        return model.DataArray(data_arrays, metadata)
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+        """
+        remaining_t = self.estimate_total_duration()
+        acquired_data = []
+        # iterate through streams
+        for stream in self._streams:
+            zstack = []
+            # for each stream, iterate through zlevels (if exist)
+            for i, z in enumerate(self._zlevels[stream]):
+                # move the focuser
+                self._actuator_f = stream.focuser.moveAbs({"z": z})
+                self._actuator_f.result()
+
+                # check if cancellation happened while the actuator future is working
+                if self._future_state == CANCELLED:
+                    raise CancelledError()
+
+                # if FM stream, then subtract one zstep time
+                if isinstance(stream, FluoStream):
+                    if i != len(self._zpos) - 1:
+                        remaining_t -= self._zstep_duration
+                        self._main_future.set_end_time(time.time() + remaining_t)
+
+                try:
+                    # acquire this single stream, and get the data
+                    self._single_acqui_f = acquire([stream], self._settings_obs)
+                    data, exp = self._single_acqui_f.result()
+
+                    # check if cancellation happened while the acquiring future is working
+                    if self._future_state == CANCELLED:
+                        raise CancelledError()
+
+                    # check on the acquired data
+                    if not data:
+                        logging.warning(
+                            "The acquired data array for stream %s is empty", stream
+                        )
+                    else:
+                        logging.info("The acquisition for stream %s is done", stream)
+
+                except CancelledError:
+                    logging.exception("The acquisition was canceled")
+                except Exception:
+                    logging.exception("The acquisition failed")
+                finally:
+                    zstack.append(data[0])
+                    # update the remaining time
+                    remaining_t -= stream.estimateAcquisitionTime()
+                    self._main_future.set_end_time(time.time() + remaining_t)
+
+            if isinstance(stream, FluoStream):
+                zcube = self._construct_zcube(zstack)
+                acquired_data.append(zcube)
+            elif isinstance(stream, EMStream):
+                acquired_data.append(zstack[0])
```

Comment:
That's way too specific. It should handle every type of stream the same way. If it has more than one zlevel -> construct Z cube.